streaming/storage/deque.go

```diff
@@ -0,0 +1,392 @@
+	tx StateTransaction
+
+	initializedFront bool
+	initializedBack  bool
+
+	firstFreeFrontSpot int
+	firstFreeBackSpot  int
+}
+
+type DequeIterator struct {
+	it Iterator
+}
+
+func NewDeque(tx StateTransaction) *Deque {
+	return &Deque{
+		tx:                 tx,
+		initializedFront:   false,
+		initializedBack:    false,
+		firstFreeFrontSpot: defaultFirstFreeFrontSpot,
+		firstFreeBackSpot:  defaultFirstFreeBackSpot,
+	}
+}
+
+func NewDequeIterator(it Iterator) *DequeIterator {
+	return &DequeIterator{
+		it: it,
+	}
+}
+
+//Adds new element to the front of the queue
+func (dq *Deque) PushFront(value proto.Message) error {
+	if !dq.initializedFront {
+		err := dq.initializeFront()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in queue.PushFront")
+		}
+	}
+
+	data, err := proto.Marshal(value)
+	if err != nil {
+		return errors.Wrap(err, "couldn't serialize given value")
+	}
+
+	return dq.appendBytes(data, dq.firstFreeFrontSpot, dq.decreaseFirstFreeFrontSpot)
+}
+
+//Adds new element to the back of the queue
+func (dq *Deque) PushBack(value proto.Message) error {
+	if !dq.initializedBack {
+		err := dq.initializeBack()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in queue.PushBack")
+		}
+	}
+
+	data, err := proto.Marshal(value)
+	if err != nil {
+		return errors.Wrap(err, "couldn't serialize given value")
+	}
+
+	return dq.appendBytes(data, dq.firstFreeBackSpot, dq.increaseFirstFreeBackSpot)
+}
+
+//This is an auxiliary function used in PushBack and PushFront. The modifier is needed,
+//to move the firstFreeFrontSpot to the left (PushFront) or the firstFreeBackSpot to the
+// right (PushBack). These functions call appendBytes with modifier set to
+// dq.decreaseFirstFreeFrontSpot and dq.increaseFirstFreeBackSpot respectively.
+func (dq *Deque) appendBytes(data []byte, index int, modifier func() error) error {
+	byteKey := getIndexKey(index)
+
+	err := dq.tx.Set(byteKey, data)
+	if err != nil {
+		return errors.Wrap(err, "couldn't add the element to the queue")
+	}
+
+	err = modifier() //modifier is a function of the dq so it modifies dq
+	if err != nil {
+		return errors.Wrap(err, "failed to actualize queue attributes")
+	}
+
+	return nil
+}
+
+//Removes the first element from the queue and returns it
+//Returns ErrNotFound if the list is empty
+func (dq *Deque) PopFront(value proto.Message) error {
+	//there is no need to call initialize here, since PeekFront calls initialize
+	err := dq.PeekFront(value)
+	if err != nil {
+		return err
+	}
+
+	return dq.pop(value, dq.firstFreeFrontSpot+1, dq.increaseFirstFreeFrontSpot)
+}
+
+//Removes the last element from the queue and returns it
+//Returns ErrNotFound if the list is empty
+func (dq *Deque) PopBack(value proto.Message) error {
+	//there is no need to call initialize here, since PeekBack calls initialize
+	err := dq.PeekBack(value)
+	if err != nil {
+		return err
+	}
+
+	return dq.pop(value, dq.firstFreeBackSpot-1, dq.decreaseFirstFreeBackSpot)
```

Comment:
I also think that those modifier shenanigans are kinda complex, and you could just call dq.decreaseFirstFreeBackSpot here, after using pop. It would probably work the same way, and would be much less complicated.

Reply:
True - I did just that